src/frequenz/sdk/timeseries/logical_meter/_formula_generators.py

```diff
@@ -0,0 +1,93 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Generators for formulas from component graphs."""
+
+from frequenz.channels import Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid import ComponentGraph
+from ...microgrid.component import ComponentCategory, ComponentMetricId
+from ._formula_engine import FormulaEngine
+from ._resampled_formula_builder import ResampledFormulaBuilder
+
+
+class FormulaGenerator:
+    """A class for generating formulas from the component graph."""
+
+    def __init__(
+        self,
+        namespace: str,
+        channel_registry: ChannelRegistry,
+        resampler_subscription_sender: Sender[ComponentMetricRequest],
+    ) -> None:
+        """Create a `FormulaGenerator` instance.
+
+        Args:
+            namespace: A namespace to use with the data-pipeline.
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
+        """
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._namespace = namespace
+
+    async def grid_power_formula(
+        self,
+        component_graph: ComponentGraph,
+    ) -> FormulaEngine:
+        """Generate a formula for calculating grid power from the component graph.
+
+        Args:
+            component_graph: The component graph for the microgrid.
+
+        Returns:
+            A formula engine that will calculate grid power values.
+
+        Raises:
+            RuntimeError: when the component graph doesn't have a `GRID` component.
+        """
+        grid_component = next(
+            (
+                comp
+                for comp in component_graph.components()
+                if comp.category == ComponentCategory.GRID
+            ),
+            None,
+        )
+
+        if grid_component is None:
+            raise RuntimeError(
+                "Unable to find a GRID component from the component graph."
+            )
+
+        grid_successors = component_graph.successors(grid_component.component_id)
```

Comment:
But if umg_meter will be by load side, then grid_successor will have only one element: JUNCTION. And then formula below won't work.